src/player.cpp

```diff
@@ -10050,74 +10041,47 @@ void player::consume_effects(item *eaten, const it_comest *comest, bool rotten)
-        // Thanks for the warning, i2amroy.
-        if ((rotten) && !(has_trait("SAPROPHAGE")) ) {
-            int nut = (rng(0, nutrition_for(comest)) * 0.66 );
-            int que = (comest->quench) * 0.66;
-            hunger -= nut;
-            thirst -= que;
-            stomach_food += nut;
-            stomach_water += que;
-            if (!has_trait("SAPROVORE") && !has_bionic("bio_digestion")) {
-                mod_healthy_mod(-30);
-            }
-        } else {
-        // Shorter GI tract, so less nutrients captured.
-            int giz_nutr = (nutrition_for(comest)) * 0.66;
-            int giz_quench = (comest->quench) * 0.66;
-            int giz_healthy = (comest->healthy) * 0.66;
-            hunger -= (giz_nutr);
-            thirst -= (giz_quench);
-            mod_healthy_mod(giz_healthy);
-            stomach_food += (giz_nutr);
-            stomach_water += (giz_quench);
-        }
-    } else if (has_trait("CARNIVORE") && (eaten->made_of("veggy") || eaten->made_of("fruit") || eaten->made_of("wheat")) &&
-      (eaten->made_of("flesh") || eaten->made_of("hflesh") || eaten->made_of("iflesh") || eaten->made_of("milk") ||
-      eaten->made_of("egg")) ) {
-          // Carnivore is stripping the good stuff out of that plant crap it's mixed up with.
-          // They WILL eat the Meat Pizza.
-          int carn_nutr = (nutrition_for(comest)) * 0.5;
-          int carn_quench = (comest->quench) * 0.5;
-          int carn_healthy = (comest->healthy) * 0.5;
-          hunger -= (carn_nutr);
-          thirst -= (carn_quench);
-          mod_healthy_mod(carn_healthy);
-          stomach_food += (carn_nutr);
-          stomach_water += (carn_quench);
-          add_msg_if_player(m_good, _("You eat the good parts and leave that indigestible plant stuff behind."));
-    } else if (has_trait("CARNIVORE") && ((eaten->made_of("flesh") || eaten->made_of("hflesh") ||
-      eaten->made_of("iflesh") || eaten->made_of("egg"))) ) {
-          // Carnivores, being specialized to consume meat, get more nutrients from a wholly-meat or egg meal.
-          if (comest->healthy < 1) {
-              int carn_healthy = (comest->healthy) + 1;
-              mod_healthy_mod(carn_healthy);
-          }
-          hunger -= nutrition_for(comest);
-          thirst -= comest->quench;
-          mod_healthy_mod(comest->healthy);
-          stomach_food += nutrition_for(comest);
-          stomach_water += comest->quench;
-    } else {
-    // Saprophages get the same boost from rotten food that others get from fresh.
-        hunger -= nutrition_for(comest);
-        thirst -= comest->quench;
-        mod_healthy_mod(comest->healthy);
-        stomach_food += nutrition_for(comest);
-        stomach_water += comest->quench;
     }
 
+    // Bio-digestion gives extra nutrition
     if (has_bionic("bio_digestion")) {
-        hunger -= rng(0, nutrition_for(comest));
+        hunger_factor += rng_float(0, 1);
```

Comment:
Realistically it seems strange to me to be able to eat something, immediately vomit it back up before you can digest anything, and still somehow have a net gain in nutrition. This was aimed at stopping that, so it's intended.